File: openclaw-plugins/video-http-tools/src/tools/stats.ts
import { Type, getConfig, httpJson, toolTextResult } from "../shared";

const WriteItemSchema = Type.Object({
  platform: Type.Optional(Type.String({ description: "平台，如 douyin，默认 douyin" })),
  title: Type.String({ description: "视频标题" }),
  publish_time: Type.String({ description: "发布时间，ISO 8601，如 2026-03-10T12:00:00Z" }),
  duration_sec: Type.Optional(Type.Number({ description: "视频时长（秒）" })),
  play_count: Type.Optional(Type.Integer({ description: "播放量" })),
  like_count: Type.Optional(Type.Integer({ description: "点赞量" })),
  comment_count: Type.Optional(Type.Integer({ description: "评论量" })),
  share_count: Type.Optional(Type.Integer({ description: "分享量" })),
  collect_count: Type.Optional(Type.Integer({ description: "收藏量" })),
  danmaku_count: Type.Optional(Type.Integer({ description: "弹幕量" })),
  completion_rate: Type.Optional(Type.Number({ description: "完播率，0.0~1.0" })),
  two_sec_exit_rate: Type.Optional(Type.Number({ description: "2s 跳出率，0.0~1.0" })),
  follower_gain: Type.Optional(Type.Integer({ description: "吸粉量" })),
  follower_loss: Type.Optional(Type.Integer({ description: "脱粉量" })),
  fan_play_ratio: Type.Optional(Type.Number({ description: "粉丝播放占比，0.0~1.0" }))
});

export default function registerStatsTools(api: any) {
  api.registerTool({
    name: "stats_write",
    description: "Write one or more video metric snapshots to the database. Each call adds a new snapshot without overwriting history.",
    parameters: Type.Object({
      items: Type.Array(WriteItemSchema, { description: "快照列表，单条也需用数组包装" })
    }),
    async execute(_id: string, params: any) {
      const cfg = getConfig(api);
      const result = await httpJson(cfg, "/stats/write", "POST", params);
      return toolTextResult(result);
    }
  });

  api.registerTool({
    name: "stats_query",
    description: "Query video metric snapshots with filters and pagination. Returns total_count for pagination. Use latest_only=true to get only the most recent snapshot per video.",
    parameters: Type.Object({
      platform: Type.Optional(Type.String({ description: "按平台过滤" })),
      title: Type.Optional(Type.String({ description: "标题模糊匹配" })),
      publish_since: Type.Optional(Type.String({ description: "发布时间起点，ISO 8601" })),
      publish_until: Type.Optional(Type.String({ description: "发布时间终点，ISO 8601" })),
      snapshot_since: Type.Optional(Type.String({ description: "快照时间起点，ISO 8601" })),
      snapshot_until: Type.Optional(Type.String({ description: "快照时间终点，ISO 8601" })),
      latest_only: Type.Optional(Type.Boolean({ description: "true 时每个视频只返回最新快照" })),
      limit: Type.Optional(Type.Integer({ minimum: 1, maximum: 500, description: "每页条数，默认 50" })),
      offset: Type.Optional(Type.Integer({ minimum: 0, description: "跳过条数，默认 0" }))
    }),
    async execute(_id: string, params: any) {
      const cfg = getConfig(api);
      const query = new URLSearchParams();
      for (const [k, v] of Object.entries(params)) {
        if (v !== undefined && v !== null) query.set(k, String(v));
      }
      const path = `/stats/query${query.toString() ? "?" + query.toString() : ""}`;
      const result = await httpJson(cfg, path, "GET");
      return toolTextResult(result);
    }
  });
}

File: agent-service/app/schemas/stats.py
"""
视频数据统计模块请求与响应 Schema。

对应接口：
  POST /stats/write
  GET  /stats/query
"""
from __future__ import annotations

from typing import Any, Dict, List, Optional

from pydantic import BaseModel, Field


# ---------------------------------------------------------------------------
# POST /stats/write
# ---------------------------------------------------------------------------


class VideoStatsWriteItem(BaseModel):
    platform: str = Field("douyin", description="平台，如 douyin")
    title: str = Field(..., description="视频标题")
    publish_time: str = Field(..., description="发布时间，ISO 8601，如 2026-03-10T12:00:00Z")
    duration_sec: Optional[float] = Field(None, description="视频时长（秒）")
    play_count: Optional[int] = Field(None, description="播放量")
    like_count: Optional[int] = Field(None, description="点赞量")
    comment_count: Optional[int] = Field(None, description="评论量")
    share_count: Optional[int] = Field(None, description="分享量")
    collect_count: Optional[int] = Field(None, description="收藏量")
    danmaku_count: Optional[int] = Field(None, description="弹幕量")
    completion_rate: Optional[float] = Field(None, description="完播率，0.0~1.0")
    two_sec_exit_rate: Optional[float] = Field(None, description="2s 跳出率，0.0~1.0")
    follower_gain: Optional[int] = Field(None, description="吸粉量")
    follower_loss: Optional[int] = Field(None, description="脱粉量")
    fan_play_ratio: Optional[float] = Field(None, description="粉丝播放占比，0.0~1.0")
    extra: Optional[Dict[str, Any]] = Field(None, description="预留扩展字段")


class StatsWriteRequest(BaseModel):
    items: List[VideoStatsWriteItem] = Field(..., description="快照列表，单条也用列表包装")


class StatsWriteResultItem(BaseModel):
    video_id: str
    snapshot_id: int


class StatsWriteData(BaseModel):
    written: int
    results: List[StatsWriteResultItem]


# ---------------------------------------------------------------------------
# GET /stats/query
# ---------------------------------------------------------------------------


class VideoStatsItem(BaseModel):
    video_id: str
    platform: str
    title: str
    publish_time: str
    duration_sec: Optional[float] = None
    snapshot_id: int
    snapshot_at: str
    play_count: Optional[int] = None
    like_count: Optional[int] = None
    comment_count: Optional[int] = None
    share_count: Optional[int] = None
    collect_count: Optional[int] = None
    danmaku_count: Optional[int] = None
    completion_rate: Optional[float] = None
    two_sec_exit_rate: Optional[float] = None
    follower_gain: Optional[int] = None
    follower_loss: Optional[int] = None
    fan_play_ratio: Optional[float] = None
    extra: Optional[str] = None


class StatsQueryData(BaseModel):
    total_count: int
    limit: int
    offset: int
    items: List[VideoStatsItem]
